### ISSUE TITLE
crash from failed assertion in networking.cpp

### ISSUE THREAD

**Issue by bpo** [JohnSully/KeyDB#55]
Hi there, in case it's useful here's a crash from a KeyDB instance built off the `RELEASE_5` branch last week, using 6 threads, virtual network interface has 8 queues. I can provide more details and/or sanitize the INFO/CLIENT payloads if it'd be helpful.

```
Jul 14 03:20:04 keydb-server[2881]: === KEYDB BUG REPORT START: Cut & paste starting from here ===
Jul 14 03:20:04 keydb-server[2881]: 2881:M 14 Jul 2019 03:20:04.799 # === ASSERTION FAILED CLIENT CONTEXT ===
Jul 14 03:20:04 keydb-server[2881]: 2881:M 14 Jul 2019 03:20:04.799 # client->flags = 2097152
Jul 14 03:20:04 keydb-server[2881]: 2881:M 14 Jul 2019 03:20:04.799 # client->fd = 425
Jul 14 03:20:04 keydb-server[2881]: 2881:M 14 Jul 2019 03:20:04.799 # client->argc = 2
Jul 14 03:20:04 keydb-server[2881]: 2881:M 14 Jul 2019 03:20:04.799 # client->argv[0] = ">" (refcount: 2)
Jul 14 03:20:04 keydb-server[2881]: 2881:M 14 Jul 2019 03:20:04.799 # client->argv[1] = "queue:keydb2" (refcount: 1)
Jul 14 03:20:04 keydb-server[2881]: 2881:M 14 Jul 2019 03:20:04.799 # === ASSERTION FAILED ===
Jul 14 03:20:04 keydb-server[2881]: 2881:M 14 Jul 2019 03:20:04.799 # ==> networking.cpp:2633 'c->cmd != NULL' is not true
Jul 14 03:20:04 keydb-server[2881]: 2881:M 14 Jul 2019 03:20:04.799 # (forcing SIGSEGV to print the bug report.)
Jul 14 03:20:04 keydb-server[2881]: 2881:M 14 Jul 2019 03:20:04.799 # KeyDB 5.0.0 crashed by signal: 11
Jul 14 03:20:04 keydb-server[2881]: 2881:M 14 Jul 2019 03:20:04.799 # Crashed running the instruction at: 0x482738
Jul 14 03:20:04 keydb-server[2881]: 2881:M 14 Jul 2019 03:20:04.799 # Accessing address: 0xffffffffffffffff
Jul 14 03:20:04 keydb-server[2881]: 2881:M 14 Jul 2019 03:20:04.799 # Failed assertion: c->cmd != NULL (networking.cpp:2633)
Jul 14 03:20:04 keydb-server[2881]: ------ STACK TRACE ------
Jul 14 03:20:04 keydb-server[2881]: EIP:
Jul 14 03:20:04 keydb-server[2881]: /usr/local/bin/keydb-server *:11042(_serverAssert+0x78) [0x482738]
Jul 14 03:20:04 keydb-server[2881]: Backtrace:
Jul 14 03:20:04 keydb-server[2881]: /usr/local/bin/keydb-server *:11042(logStackTrace(ucontext_t*)+0x39) [0x4851a9]
Jul 14 03:20:04 keydb-server[2881]: /usr/local/bin/keydb-server *:11042(sigsegvHandler(int, siginfo_t*, void*)+0xa4) [0x485824]
Jul 14 03:20:04 keydb-server[2881]: /lib64/libpthread.so.0(+0x121f0) [0x7fc293da61f0]
Jul 14 03:20:04 keydb-server[2881]: /usr/local/bin/keydb-server *:11042(_serverAssert+0x78) [0x482738]
Jul 14 03:20:04 keydb-server[2881]: /usr/local/bin/keydb-server *:11042(rewriteClientCommandVector(client*, int, ...)+0x120) [0x447df0]
Jul 14 03:20:04 keydb-server[2881]: /usr/local/bin/keydb-server *:11042(blockingPopGenericCommand(client*, int)+0x1db) [0x4666bb]
Jul 14 03:20:04 keydb-server[2881]: /usr/local/bin/keydb-server *:11042(call(client*, int)+0xbd) [0x43a6dd]
Jul 14 03:20:04 keydb-server[2881]: /usr/local/bin/keydb-server *:11042(processCommand(client*, int)+0x789) [0x43ebc9]
Jul 14 03:20:04 keydb-server[2881]: /usr/local/bin/keydb-server *:11042(processCommandAndResetClient(client*, int)+0x21) [0x447391]
Jul 14 03:20:04 keydb-server[2881]: /usr/local/bin/keydb-server *:11042(processInputBuffer(client*, int)+0x7a) [0x44bbba]
Jul 14 03:20:04 keydb-server[2881]: /usr/local/bin/keydb-server *:11042(readQueryFromClient(aeEventLoop*, int, void*, int)+0x338) [0x44c138]
Jul 14 03:20:04 keydb-server[2881]: /usr/local/bin/keydb-server *:11042(ProcessEventCore+0xe9) [0x433409]
Jul 14 03:20:04 keydb-server[2881]: /usr/local/bin/keydb-server *:11042(aeProcessEvents+0x14d) [0x43369d]
Jul 14 03:20:04 keydb-server[2881]: /usr/local/bin/keydb-server *:11042(aeMain+0x2d) [0x433a0d]
Jul 14 03:20:04 keydb-server[2881]: /usr/local/bin/keydb-server *:11042(workerThreadMain(void*)+0x90) [0x437aa0]
Jul 14 03:20:04 keydb-server[2881]: /lib64/libpthread.so.0(+0x754b) [0x7fc293d9b54b]
Jul 14 03:20:04 keydb-server[2881]: /lib64/libc.so.6(clone+0x3f) [0x7fc293ad32ff]
```

**Comment by JohnSully** [comment]
I assume this is intermittent?

I see blockingPopGenericCommand in the callstack.  Can you give me an idea of what other commands may be in flight around the same time?  I will setup a torture test to suss it out.

Also FYI RELEASE_5 isn't officially released yet so I've been updating it silently with bugfixes (though none that I think would resolve this).  At this point I will only add bug fixes not new features so I do appreciate you testing it.

**Comment by JohnSully** [comment]
Do you have any modules loaded?  

I think I may have found a code path that could cause this if modules are used.

**Comment by bpo** [comment]
> I assume this is intermittent?

Yes, one crash after ~1 day of uptime.

> Do you have any modules loaded?

No modules

> Can you give me an idea of what other commands may be in flight around the same time?

Counts by commands last executed by client:

```
   2 blpop
 224 brpop
   8 evalsha
  38 exec
  18 get
   1 replconf
  16 scard
   1 script
   2 set
   1 slowlog
   1 strlen
   2 xadd
1689 xreadgroup
   1 zrangebyscore
```

**Comment by bpo** [comment]
Same machine was showing these periodically in the logs, for a handful of `fd` values, not clear to me if it's related.

> epoll_ctl failed: No such file or directory
> 2881:M 14 Jul 2019 02:33:43.201 # Error registering fd event for the new client: No such file or directory (fd=290)

Reminiscent of https://github.com/antirez/redis/issues/2479, but Sentinel is uninvolved and we did not see any connectivity issues on the replica (prior to the master crash that is)

**Comment by JohnSully** [comment]
It's tricky because argv[0] should be the hardcoded value "rpop" but instead is ">".  Reading the code this should be impossible.

It's set to that value not far away in blockingPopGenericCommand:
rewriteClientCommandVector(c,2,
                        (where == LIST_HEAD) ? shared.lpop : shared.rpop,
                        c->argv[j]);

I have a few boxes dedicated to fuzzing, I'll focus them on brpop.

**Comment by JohnSully** [comment]
This is fixed in change aae0fdc11

The root cause is that shared objects such as shared.rpop aren't marked as shared.  So the refcount is being incremented and decremented.  This isn't a problem with single threading however in a multithreaded environment client arguments must either be local to the current thread or shared (because the refcount is modified outside the lock).

Eventually shared.rpop's refcount hits zero due to the synchronization error and is freed.  At some later point malloc recycles the memory and a new object is placed there.  At that point the command lookup will fail and this assert will fire.

The fix is to correctly mark these and all other similar objects as shared.
I will continue to run my stress test for a few days to further validate the fix.
